File: src/main/java/com/hazza/leetcode/SegmentOne/No22GenerateParentheses.java
package com.hazza.leetcode.SegmentOne;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * https://leetcode.com/problems/generate-parentheses/#/description
 */
public class No22GenerateParentheses {
    //solution 1
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        if (n < 1) return res;

        insertBrackets(res, "", 0, 0, n);

        return res;
    }

    private void insertBrackets(List<String> res, String str, int openBrackets, int closeBrackets, int max) {
        if (openBrackets + closeBrackets == max * 2) res.add(str);

        if (openBrackets < max) insertBrackets(res, str + '(', openBrackets + 1, closeBrackets, max);
        if (closeBrackets < openBrackets) insertBrackets(res, str + ')', openBrackets, closeBrackets + 1, max);
    }

    //solution 2
    public List<String> generateParenthesis2(int n) {
        List<List<String>> lists = new ArrayList<>();
        lists.add(Collections.singletonList(""));

        for (int i = 1; i <= n; ++i)
        {
            final List<String> list = new ArrayList<>();
            for (int j = 0; j < i; ++j)
            {
                for (final String first: lists.get(j))
                {
                    for (final String second: lists.get(i - 1 - j))
                    {
                        list.add("(" + first + ")" + second);
                    }
                }
            }
            lists.add(list);
        }
        return lists.get(lists.size() - 1);
    }

}

File: src/main/python/leetcode_by_python/SegmentOne/No22_Generate_Parentheses.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/18 20:53
# https://leetcode.com/problems/generate-parentheses/
from typing import List


class Solution:
    # solution 1
    def generateParenthesis1(self, n: int) -> List[str]:
        def insertBrackets(leftBracketsNum, rightBracketsNum, n, str, res):
            if leftBracketsNum + rightBracketsNum == 2 * n:
                res.append(str)
            else:
                if leftBracketsNum < n:
                    insertBrackets(leftBracketsNum + 1, rightBracketsNum, n, str + '(', res)
                if rightBracketsNum < leftBracketsNum:
                    insertBrackets(leftBracketsNum, rightBracketsNum + 1, n, str + ')', res)

        res = []
        insertBrackets(0, 0, n, '', res)

        return res

    #solution 2
    def generateParenthesis2(self, n: int) -> List[str]:
        res = [['']]
        for i in range(1, n + 1):
            res.append(['(' + first + ')' + second for j in range(i) for first in res[j] for second in res[i - j - 1]])

        return res[-1]


if __name__ == '__main__':
    solution = Solution()
    assert sorted(solution.generateParenthesis1(3)) == sorted(["((()))", "(()())", "(())()", "()(())", "()()()"])
    assert sorted(solution.generateParenthesis2(3)) == sorted(["((()))", "(()())", "(())()", "()(())", "()()()"])
    assert sorted(solution.generateParenthesis2(10)) == sorted(solution.generateParenthesis2(10))